### PR TITLE
[avahi] bug fix and small enhancements for avahi publisher

### PR DESCRIPTION
**Pull request by superwhd**
This PR contains a bug fix and some small enhancements for the Avahi publisher.

Bug fix: 
- Checking the address size before de-referencing the address pointer.

Enhancements:
- Don't free the `mRecordBrowser` manually in the `HandleResolveResult` callback. The `mRecordBrowser` will be freed when the subscription is released.
- Log the error message instead of error code when there's an Avahi error.
- Revised the cases where we fetch an Avahi error. There used to be some inappropriate places we fetch an error code from the Avahi client, which may produce false alarm logs.

### CODE REVIEW

File: src/mdns/mdns_avahi.cpp [hunk 1]
```diff
@@ -1086,7 +1086,7 @@ void PublisherAvahi::ServiceSubscription::HandleResolveResult(AvahiServiceResolv
         // NOTE: This `ServiceSubscrption` object may be freed in `OnServiceResolved`.
         mPublisherAvahi->OnServiceResolved(mType, instanceInfo);
     }
-    else if (avahi_client_errno(mPublisherAvahi->mClient) != AVAHI_OK)
+    else if (avahiError)
```
Comment (simonlingoogle) [inline src/mdns/mdns_avahi.cpp:1089]:
```suggestion
    else if (avahiError != AVAHI_OK)
```

File: src/mdns/mdns_avahi.cpp [hunk 2]
```diff
@@ -1038,30 +1038,30 @@ void PublisherAvahi::ServiceSubscription::HandleResolveResult(AvahiServiceResolv
     Ip6Address             address;
     size_t                 totalTxtSize = 0;
     DiscoveredInstanceInfo instanceInfo;
-    bool                   resolved = false;
+    bool                   resolved   = false;
+    int                    avahiError = 0;
```
Comment (simonlingoogle) [inline src/mdns/mdns_avahi.cpp:1042]:
```suggestion
    int                    avahiError = AVAHI_OK;
```

File: src/mdns/mdns_avahi.cpp [hunk 3]
```diff
@@ -1167,17 +1167,16 @@ void PublisherAvahi::HostSubscription::HandleResolveResult(AvahiRecordBrowser *
     OTBR_UNUSED_VARIABLE(aFlags);
 
     Ip6Address address;
-    bool       resolved = false;
+    bool       resolved   = false;
+    int        avahiError = 0;
```
Comment (simonlingoogle) [inline src/mdns/mdns_avahi.cpp:1171]:
```suggestion
    int        avahiError = AVAHI_OK;
```

File: src/mdns/mdns_avahi.cpp [hunk 4]
```diff
@@ -1167,17 +1167,16 @@ void PublisherAvahi::HostSubscription::HandleResolveResult(AvahiRecordBrowser *
...
 
     VerifyOrExit(aEvent == AVAHI_BROWSER_NEW);
+    VerifyOrExit(aSize == 16, otbrLogErr("unexpected address data length: %u", aSize),
+                 avahiError = AVAHI_ERR_INVALID_ADDRESS);
```
Comment (simonlingoogle) [inline src/mdns/mdns_avahi.cpp:1175]:
```suggestion
    VerifyOrExit(aSize == 16 || aSize == 4, otbrLogErr("unexpected address data length: %u", aSize),
                 avahiError = AVAHI_ERR_INVALID_ADDRESS);
    VerifyOrExit(aSize == 16, otbrLogErr("IPv4 address ignored"),
                 avahiError = AVAHI_ERR_INVALID_ADDRESS);
```
nit: may have better log for IPv4 address

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1: 

with nits

File: src/mdns/mdns_avahi.cpp [hunk 5]
```diff
@@ -1167,17 +1167,17 @@ void PublisherAvahi::HostSubscription::HandleResolveResult(AvahiRecordBrowser *
     OTBR_UNUSED_VARIABLE(aFlags);
 
     Ip6Address address;
-    bool       resolved = false;
+    bool       resolved   = false;
+    int        avahiError = AVAHI_OK;
 
     VerifyOrExit(aEvent == AVAHI_BROWSER_NEW);
+    VerifyOrExit(aSize == 16 || aSize == 4, otbrLogErr("unexpected address data length: %u", aSize),
```
Comment (jwhui) [inline src/mdns/mdns_avahi.cpp:1174]: Can we use `sizeof()` instead of magic numbers here? If not `sizeof()`, then defined constants?

**Review by jwhui — APPROVED** [review]

**Pull request merged**